### DISCUSSION TITLE
jsonnet as first class citizen

### DISCUSSION THREAD

**Discussion by schmurfy** [concourse/concourse#6260] [Ideas]
I looked briefly at yaml but quickly switched to jsonnet for a lot of reasons amongst which:
- functions allow code reuse
- built time variables help avoiding repetition
- import allow splitting a large pipeline in multiple reusable parts

dynamic `across` is a step in the right direction but would not allow me by itself
To handle pull request what I am currently doing is splitting my code in two pipelines:
- the first one just listen on a github-pr resource (and git resource for releases but let's ignore that here)
- when a change is detected a task check what has changed (we have multiple projects in the same repo)
- jsonnet is used to build the pipeline which takes as input the list of apps to build & test and the result is then passed to `fly set-pipeline` to spawn the new pipeline for handling only this pull request
- when this pipeline completed successfully it execute a task to auto archive itself

The idea was to have one pipeline handling one PR (and its changes) so you have a clear, uncluttered view, once the pr is validated the pipeline goes away but since it is archived you can still access it using the links on the pr itself (or by showin g archived pipelines of course)

I find jsonnet coupled to concourse to be such a powerful tool it would be a shame to not ease its usage to build even greater pipelines without having to resort to calling fly by hand.

Here is what a pull-request job looks, in that case it builds two applications
![screenshot_616](https://user-images.githubusercontent.com/2983/98643209-f45aef00-232e-11eb-876c-0c510ecde898.png)

PS: the need to spawn a pipeline with the list of apps to build is a direct consequence of being unable to have conditional jobs.

**Comment by chenbh** [comment]
> jsonnet is used to build the pipeline which takes as input the list of apps to build & test and the result is then passed to fly set-pipeline to spawn the new pipeline for handling only this pull request

This feels like the flow that [instance pipelines + `set_pipeline` step](https://github.com/concourse/rfcs/blob/master/034-instanced-pipelines/proposal.md#proposal) is trying to solve. Although I'm unsure how the "list of apps to build" would be handled.

What kinds of features are you thinking of to make jsonnet a first class citizen? I don't think we want to embed a templating engine inside fly (see https://github.com/concourse/concourse/pull/5936#issuecomment-667192099), and it already supports [reading from stdin](https://github.com/concourse/concourse/pull/5861) so you can do `jsonnet ./template | fly set-pipeline --config -`

Side note, what do you think of jsonnet vs [ytt](https://get-ytt.io/)? I haven't used either much and I'm curious on your opinion.

**Reply by schmurfy** [reply]
I used both ytt and jsonnet and while both have similar features I find jsonnet to be easier to use and it also produces easier to read code, ytt does not support `import` for example, you can feed multiple files to ytt but a file cannot import another.

As for jsonnet your are right fly does allow taking generated files and that's how my current pipelines work but I was thinking about the dynamic `accross` which would be the missing piece to recreate what I have but it will only support loading a json/yaml pipeline.
That said maybe an intermediate task could be used to generate the jsonnet files.

I did had a clear idea on what could be done to promote jsonnet, I just wanted to start a discussion and see what comes out of it :)

I am curious to hear if anyone else uses it.
